### DISCUSSION TITLE
Theia Source code not running on vscode

### DISCUSSION THREAD

**Discussion by Ahmed-Bilal-01** [eclipse-theia/theia#13431] [General]
after I run yarn run test:playwright  it gives following error:
![theia issue](https://github.com/eclipse-theia/theia/assets/137169507/7a828c8b-d535-47c5-8d0b-9ee437daf93d)

**Comment by JonasHelming** [comment]
Hi, what operating system are you using?

**Reply by msujew** [reply]
It's Windows (see file paths). I've also experienced issues running the playwright tests on Windows. In this case, it looks like the command is started with arguments i.e. `THEIA_CONFIG_DIR=...`, which isn't supported on Windows.

**Reply by Ahmed-Bilal-01** [reply]
Yes, I am using Windows . So how can it be resolved?

**Reply by msujew** [reply]
The problematic line is [here](https://github.com/eclipse-theia/theia/blob/257cef2d690bd4cb09c8991d2ee0fa79bf622d32/examples/playwright/package.json#L18). I'm not sure you can successfully run the line (due to the `$PWD`) on Windows, even if the env variable can be set using something like `cross-env`. @planger Any input?

**Reply by planger** [reply]
Yes, these scripts are mainly targeted at the CI environment running on Linux. As @msujew points out, the problematic parts are setting the env variables to set a ephemeral Theia config directory. We should generalize those yarn scripts at some point, e.g. using cross-env or similar.

I don't have a Windows machine at hand to test this right now, but as a workaround for you @Ahmed-Bilal-01 it should work to do something like:
```json
   "theia:start": "rimraf .tmp.cfg && set THEIA_CONFIG_DIR=./.tmp.cfg && yarn --cwd ../browser start"
```
This uses `set` to define environment variables on Windows. If you encounter issues with setting the environment variable inline, we could consider using a cross-platform package like `cross-env` for setting environment variables.
First would we need to add it as a dev dependency:
```bash
yarn add --dev cross-env
```
Then we could use it in the yarn script:
```json
"theia:start": "rimraf .tmp.cfg && cross-env THEIA_CONFIG_DIR=./.tmp.cfg yarn --cwd ../browser start"
```
Regarding the `$PWD`, I don't think it is strictly needed, but we would need to test this.

Edit: Regarding `$PWD`, actually I might be wrong, I guess Theia uses that path relative from the working directory in which Theia is started. As we start Theia with a CWD from one folder up (../browser), the path ./.tmp.cfg likely is wrong and would result into setting up the temporary config dir in ../browser/.tmp.cfg.
However, if it is just about running it locally, please just check where the .tmp.cfg folder is created and delete it before running the tests. This is just to avoid polluting your user Theia config dir and spilling config data from one test run into the next one.

**Reply by Ahmed-Bilal-01** [reply]
![thia issue 2](https://github.com/eclipse-theia/theia/assets/137169507/cc68aa57-7922-4ba0-9ab1-34872ea12a7d)
I changed  ../browser start with build, it build successfully but when I again ran yarn run test:playwright  it gave same Theia_Config_Dir issue which i posted before

**Reply by planger** [reply]
I'm sorry you are still having troubles getting this to work. It is difficult for me to debug this without seeing all changes and knowing which exact script you ran from which CWD.

But from what I can see in the screenshot:
Are you sure you ran the `theia:start` script you adapted regarding the env variable in the root `package.json` and not the one (which you potentially didn't adapt accordingly) in `/examples/playwright/package.json`? Because the `theia:start` script you've added to the root looks suspicious:, it invokes `yarn --cwd ../browser build` in the end. The `cwd` seems wrong given it is in the root package.json, also why is it calling `build` and not `start`?

But I might miss something else here from this limited view on the overall configuration on your end.

**Comment by Ahmed-Bilal-01** [comment]
@planger you are right, I was targeting wrong package.json (root). Now i did change in playwright package.json file, I replaced:
 "theia:start": "rimraf .tmp.cfg && set THEIA_CONFIG_DIR=./.tmp.cfg && yarn --cwd ../browser start"
   with
"theia:start": "rimraf .tmp.cfg && cross-env THEIA_CONFIG_DIR=./.tmp.cfg yarn --cwd ../browser start"

Now it gave a different kind of error:
![theia issue 5](https://github.com/eclipse-theia/theia/assets/137169507/91ee0615-76d5-49d5-9746-ed77ca0ffc5d)
![theia issue 6](https://github.com/eclipse-theia/theia/assets/137169507/443e2d74-f6f6-4da5-be24-00891ed9808e)

Please guide me on this and when giving solution, its a request if you can please mention the path as well where to make change coz there are alot of files and many have same names so it becomes a bit difficult to figure out the right one.

**Reply by planger** [reply]
Actually, this was successful. The Playwright tests have indeed executed successfully, there was just one test that failed.
But as I just learned from [your message below](https://github.com/eclipse-theia/theia/discussions/13431#discussioncomment-8705932) we were actually chasing the wrong problem, as your goal never was to execute the Playwright tests.

**Comment by Ahmed-Bilal-01** [comment]
@planger , Let me tell you how I started:
- cloned code from github.
- installed node, npm and yarn.
- using yarn installed all packages (yarn install --force)
- used yarn run all
- Now i am trying to open the example application on browser so running yarn run test:playwright .

let me know is it the right way or i have missed any step?

**Reply by msujew** [reply]
@Ahmed-Bilal-01 Are you just trying to build & run Theia from sources? Have you seen our [quickstart guide](https://www.tagesschau.de/wirtschaft/unternehmen/lufthansa-verdi-streik-bodenpersonal-100.html)? Playwright is only used to perform automated UI testing for Theia.

**Reply by Ahmed-Bilal-01** [reply]
Yes, I want to build and run Theia using source code.  I need help on that, which scripts are must and required for that
I think you have inserted a wrong link for quickstart guide

**Reply by planger** [reply]
To start Theia from the source code of this repository, please follow the steps documented here:
https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md#quick-start

This will start the example application ([browser](https://github.com/eclipse-theia/theia/tree/master/examples/browser) or [electron](https://github.com/eclipse-theia/theia/tree/master/examples/electron)), which is part of the Theia repository.

Usually, however, people want to build an own IDE based on Theia (and not start vanilla Theia from the Theia source code). Therefore, please have a look at https://theia-ide.org/docs/composing_applications. This will help you create your application that consumes the Theia modules as dependencies, instead of running it directly from source.

Running the example applications mentioned above from the Theia source code is usually only done by people who are contributing to the Theia sources directly.

**Reply by Ahmed-Bilal-01** [reply]
Actually I want to customize IDE for embedded systems. Specifically to burn code on TI microcontrollers. So, for that should I do changes in source code or use the other option that you said people use?
What are pros and cons of both?

**Reply by JonasHelming** [reply]
Hi,

very likely, you will not need to build from the sources, but rather extend it. You can either create a VS Code extension of a Theia extension with you custom features and then "plug" them into Theia. Consider Theia as a platform, a framework. Like React or Node.
It provides API for you to use to add your own features as well as to customize the features that are already there. The only reason to change the sources of Theia would be if you want to do something that the Theia API does not already support. However, even in this case, you typically try to contribute this adaption to the Theia project and once a new Theia version with your changes is available, you adopt Theia again as a dependency. Only in a state of your project where you expect to do frequent changes on the underlying platform Theia, it would make sense to build against the sources.
Your use case to build an IDE for embedded development is not an unusual one for Theia, several stakeholders use Theia for the same. So depending on your design, it is likely, that you can use Theia as is, just as a dependency. So, as Philip outlined, this documentation is a good start: https://theia-ide.org/docs/composing_applications

I hope this helps!

**Reply by planger** [reply]
One additional note on what Jonas said: as you are interested in creating an IDE for embedded systems, you may want to start off from [CDT Cloud Blueprint](https://github.com/eclipse-cdt-cloud/cdt-cloud-blueprint). It is template for building custom web-based C/C++ IDEs that can by using Theia extensions or VS Code extensions be enhanced for additional languages (e.g. Rust) or custom editors, e.g. for configuring specific toolchains or hardware devices. Maybe you are interested in watching an introductory talk on [building a custom IDE at the example of the Raspberry Pi Pico](https://www.youtube.com/watch?v=7h2IoWZyzVo).

**Reply by Ahmed-Bilal-01** [reply]
As per your guidance I have started working on https://theia-ide.org/docs/composing_applications but there are issues appearing in running it now. On Windows there were plenty of issues coming so along with it I tried on Ubuntu and initial steps were successfully implemented on it.
However its not starting the app.
First of all there is this error about @electron
![theia linux 1](https://github.com/eclipse-theia/theia/assets/137169507/4e865dba-55c6-4c18-a370-09a878d51717)
it says Failed to resolve module: @theia/electron but still builds successfully. 

Second when I execute yarn start or yarn start /ahmed --hostname 0.0.0.0 --port 8080, it stuck after reaching Deploy Plugins and doesn't proceed further to open app
![theia linux 2](https://github.com/eclipse-theia/theia/assets/137169507/6e557aee-cc67-4d4f-8858-4e5771dd9e8b)


I am using this as package.json
{
    "private": true,
    "dependencies": {
      "@theia/callhierarchy": "latest",
      "@theia/file-search": "latest",
      "@theia/git": "latest",
      "@theia/markers": "latest",
      "@theia/messages": "latest",
      "@theia/navigator": "latest",
      "@theia/outline-view": "latest",
      "@theia/plugin-ext-vscode": "latest",
      "@theia/preferences": "latest",
      "@theia/preview": "latest",
      "@theia/search-in-workspace": "latest",
      "@theia/terminal": "latest",
      "@theia/vsx-registry": "latest"
    },
    "devDependencies": {
      "@theia/cli": "latest"
    },
    "scripts": {
      "prepare": "yarn run clean && yarn build && yarn run download:plugins",
      "clean": "theia clean",
      "build": "theia build --mode development",
      "start": "theia start --plugins=local-dir:plugins",
      "download:plugins": "theia download:plugins"
    },
    "theiaPluginsDir": "plugins",
    "theiaPlugins": {
      "vscode-builtin-extensions-pack": "https://open-vsx.org/api/eclipse-theia/builtin-extension-pack/1.50.1/file/eclipse-theia.builtin-extension-pack-1.50.1.vsix"
    },
    "theiaPluginsExcludeIds": [
      "ms-vscode.js-debug-companion",
      "vscode.extension-editing",
      "vscode.git",
      "vscode.git-ui",
      "vscode.github",
      "vscode.github-authentication",
      "vscode.microsoft-authentication"
    ]
  }

**Reply by planger** [reply]
Actually, it looks like your build succeeded successfully and your Theia backend runs on `8080`. Now just point a browser to `localhost:8080` and you should see a Theia application being loaded in your browser.

**Reply by Ahmed-Bilal-01** [reply]
@planger Yes, thanks it worked.
But as my main target is desktop app, I am concerned why it's not running as a desktop app?

Also, I have gone through the Customized IDE for PICO video. I have few questions regarding that:
1- Compilation of c/c++ is possible on Theia as well so why to use CDT Cloud IDE and not Theia directly?
2- Actually I am working on R&D (replacing Texas Instrument's Code Composer Studio IDE) so my microcontroller is completely different (its RM57 by Texas Instrument). Main things that I have pointed out till now is, that how will I be able to integrate board with the IDE and other stuff related to burning and debugging code on microcontroller like you did for PICO (established complete communication between IDE and PICO).

I need detailed guidance on it, it will be much helpful if you share this process documentations or any helping material etc.

**Reply by msujew** [reply]
@Ahmed-Bilal-01 The commands you've executed just started the browser app. There is likely a second set of commands that builds & starts the electron app.

> I need detailed guidance on it, it will be much helpful if you share this process documentations or any helping material etc.

Given the scope of the framework we provide, it's difficult to provide/write documentation for such specific use cases (you can probably imagine that there are countless use cases for Theia). Unless you want to dig into Theia yourself, you might want to take a look at the [support page of Theia](https://theia-ide.org/support/) were you will find companies that are specialized in helping people/companies build their own Theia IDE.

**Reply by planger** [reply]
Maybe to add to it:

Are you aware of the [Theia generator](https://github.com/eclipse-theia/generator-theia-extension)? This should help to generate a working Theia application for you with both, Electron and Browser support.

If you don't already have a separate node module with a `package.json` for running your Theia app in Electron, I suggest to generate a new Theia application using the generator above to get an app that can be executed as an Electron app too. Otherwise you'd need to modify or create a separate module with a `package.json` similar to the one you have with a few modifications. See the [example electron `package.json`](https://github.com/eclipse-theia/theia/blob/master/examples/electron/package.json) as a reference.

Regarding CDT Cloud:
CDT Cloud (and CDT Cloud Blueprint) is essentially plain Theia with additional VS Code extensions (and a few Theia extensions) to make it capable of handling C/C++ development related tasks out of the box, such as vscode-clangd for C/C++ editing, cdt-gdb-adapter for enabling debugging with GDB, CMake, example project generation, etc.
Sure, you are right, you can add all that to a vanilla Theia and land at the same spot. CDT Cloud Blueprint is just meant to help you selecting the typical configuration for a C/C++ IDE based on Theia and get there faster.

Regarding the PICO talk and what's needed to implement those features:
The demo shown in the PICO talk is CDT Cloud Blueprint + some PICO specific extensions as outlined in the talk. You can also have a look at the source code of this demo: https://github.com/eclipsesource/cdt-cloud-blueprint/tree/osweek-pico
The code is not (yet) polished, but should hopefully give you an idea in addition to the talk, how the different features are implemented.

**Reply by Ahmed-Bilal-01** [reply]
Hello @planger , @msujew , I hope you guys are doing well. 

Please let me know how I can integrate my own custom compiler in Cloud CDT for compilation of C++ files? at the moment gcc is being used but I want to integrate my own custom compiler according to my requirements, so I need guidance in that.

**Reply by planger** [reply]
Hi @Ahmed-Bilal-01, it is probably best to open a new discussion topic in the [CDT Cloud forums](https://github.com/eclipse-cdt-cloud/cdt-cloud/discussions) for this particular question and discuss this in more detail there.

My initial thought to your question is: CDT Cloud (Blueprint) isn't really tied to a specific compiler. One of the example projects happens to be a CMake project that uses GCC in this configuration by default. If you want to force another compiler, you can use the `-DCMAKE_C_COMPILER` or `-CMAKE_CXX_COMPILER` command line flags of CMake in the CMake call (e.g. via `.theia/tasks.json`) or adjust the `CMakeLists.txt` accordingly (`set(CMAKE_C_COMPILER "XYZ")`).

With respect to editing support with clangd, all that matters is the generated `compile_commands.json`. If you use CMAKE this is generated automatically if you set `set(CMAKE_EXPORT_COMPILE_COMMANDS ON)` and it should correspond to the whatever compiler you've used during the compile step.
